Fixes #38879 - Update BarChart to Patternfly5

tlabaj opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

Fixes:  #38879 

This PR updates the Bar Chart to use the PF5 components.  
API's have been maintained as to not introduce breaking changes to any plugins that may be using the component.
Tests have been updated to use RTL.

Before:
<img width="632" height="359" alt="Screenshot 2025-10-29 at 4 21 00 PM" src="https://github.com/user-attachments/assets/f28be5bf-5a5e-4aaa-9ba9-73a143de7eb2" />


After:
<img width="743" height="383" alt="Screenshot 2025-10-30 at 10 31 37 AM" src="https://github.com/user-attachments/assets/edd94429-9980-4729-80e8-04b11bd11b48" />

adamruzicka commented:
Looking at those screenshots, could the chart fill the entire width of the card like it used to?

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,7 +1,12 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import { BarChart as PfBarChart } from 'patternfly-react';
-import { getBarChartConfig } from '../../../../../services/charts/BarChartService';
```
The `BarChartService.js` file could be deleted when it's not used anymore

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
Not used

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -16,24 +21,84 @@ const BarChart = ({
-  return <MessageBox msg={noDataMsg} icontype="info" />;
+
+  // Configuration based on size
+  const getChartDimensions = () => {
```
As Adam noticed, it would be nice if it filled the space

tlabaj replied:
I do not have data in my bar chart, but this is how it looks filling the space.

<img width="647" height="373" alt="image" src="https://github.com/user-attachments/assets/894e81d5-ad1a-4be7-8955-a593f1ebc49c" />

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -16,24 +21,84 @@ const BarChart = ({
+      name: item[0],
+      color: item[2] || undefined, // custom color if provided
+    }));
+  };
```
This function could be moved outside of the `BarChart` component

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -16,24 +21,84 @@ const BarChart = ({
+      x: item[0], // label
+      y: item[1], // value
+      name: item[0],
+      color: item[2] || undefined, // custom color if provided
```
The custom coloring doesn't work

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/BarChart.test.js:
```diff
@@ -1,27 +1,139 @@
-import { shallow } from 'enzyme';
 import React from 'react';
+import { render, screen, fireEvent } from '@testing-library/react';
```
`fireEvent` is not used

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/BarChart.test.js:
```diff
@@ -1,27 +1,139 @@
+  });
+
+  it('renders empty state when no data provided', () => {
+    const { container } = render(<BarChart data={emptyData} noDataMsg="No data available" />);
```
'container' is declared but its value is never read.ts(6133)

The same happens  for "renders with custom axis labels" test

kmalyjur reviewed (changes requested):
Thank you, I'm leaving some comments.
I haven't tested the chart in the foreman_openscap yet.

lhellebr commented:
@kmalyjur Is this ~finished enough to test with packit?

kmalyjur commented:
> @kmalyjur Is this ~finished enough to test with packit?

@lhellebr I'm not sure, to be honest. I would wait for the changes.

kmalyjur reviewed (changes requested):
Everything is great now, but I forgot that when deleting functions (`getBarChartConfig`) in foreman core, we need to add a deprecation message first. I'm sorry for not mentioning this before.

There is a deprecation function for that, here is an example of usage: https://github.com/theforeman/foreman/pull/8267/files.
The message could be something like "Please use COMPONENT_NAME from IMPORT_PATH instead".
Could you please not delete `getBarChartConfig` (and the tests) and deprecate it instead?

Also, the file _foreman/config/irbrc_history_ could be deleted.

kmalyjur commented:
Thank you, now it just needs to be verified

lhellebr commented:
/packit build

lhellebr commented:
https://copr.fedorainfracloud.org/coprs/fulltext/?fulltext=packit%2Ftheforeman-foreman-10743 is missing, can someone build it please?

adamruzicka commented:
/packit build

lhellebr commented:
FailedQA with foreman-3.18.0-0.20260107161624001555.pr10743.8932.g7a7c7f239.el9.noarch on Firefox 146.0 build 20251211125526 and Chromium 143.0.7499.169.

When hovered, the tooltip isn't shown so it's not at all clear from the graph what value exactly the bar represents.
Original state:
<img width="796" height="523" alt="Screenshot From 2026-01-08 15-32-37" src="https://github.com/user-attachments/assets/55d750c2-417f-4295-9d90-8eff5c02f767" />

New state:
<img width="796" height="523" alt="Screenshot From 2026-01-08 15-32-52" src="https://github.com/user-attachments/assets/8aa56176-10cf-4427-96db-2b4cfc8aa210" />

In SCAP, there is a table next to the graph, but that's not true for other places.

lhellebr commented:
@tlabaj there have been 57 commits added to this PR, that doesn't seem right

tlabaj commented:
@lhellebr @kmalyjur I updated the tooltip. I also fixed the commit issue.

<img width="980" height="441" alt="image" src="https://github.com/user-attachments/assets/7cf954ba-18b4-44f9-b19f-e36f01d36041" />

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,234 @@
+const transformData = rawData => {
+  if (!rawData || !Array.isArray(rawData)) return [];
+
+  return rawData.map(item => ({
+    x: item[0], // label
+    y: item[1], // value
+    name: item[0],
+    color: item[2] || undefined, // custom color if provided
+  }));
```
```suggestion
  return rawData.map((item, index) => ({
    x: item?.x ?? item[0],  // label
    y: item?.y ?? item[1], // value
    name: item?.name ?? (item?.x ?? item[0]),
    color: item?.color ?? (item[2] || undefined), // custom color if provided
  }));
```

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,234 @@
+        ariaDesc={title?.text || 'Bar chart'}
+        containerComponent={
+          <ChartVoronoiContainer
+            labels={({ datum }) => getTooltipValue(datum)}
```
This makes the hover area vertical slices, so it's easier to trigger the tooltip

```suggestion
          <ChartVoronoiContainer
            voronoiDimension="x"
            labels={({ datum }) => getTooltipValue(datum)}
```

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,234 @@
+            labelComponent={
+              <ChartTooltip
+                constrainToVisibleArea
+                renderInPortal
```
Deleting `renderInPortal` 

```suggestion
                constrainToVisibleArea
```

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```suggestion
BarChart.propTypes = {
  data: PropTypes.arrayOf(PropTypes.any),
```

kmalyjur reviewed (commented):
I tested this using `bundle exec rake seed:reports origin=Puppet hosts=3 reports=5 lines=10`, but I couldn't see the tooltips until I made these adjustments.
What do you think about it? If it works for you just fine without my changes, I might have something wrong with my environment. I'm honestly not sure about it.

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,234 @@
+                }}
+              />
+              <span>Value:</span>
+              <span style={{ marginLeft: 'var(--pf-v5-global--spacer--xs)' }}>
```
We should move the inline styles into a separate BarChart.scss.

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,234 @@
-        unloadBeforeLoad={unloadData}
-      />
+      <g>
+        <foreignObject
```
The tooltip looks great, however, can I please ask why `foreignObject` is used instead of `ChartTooltip` from PF?

tlabaj replied:
I am using the `ChartTooltip`.  I followed the PF [example](https://v5-archive.patternfly.org/charts/tooltips#embedded-html) on how to embed HTML in the tooltip and `foreignObject` is used.   This was the best way to get the tooltip to look like it was with the prior implementation.

kmalyjur replied:
That makes sense, thank you for the explanation

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,234 @@
+    datum: PropTypes.shape({
+      x: PropTypes.any,
+      y: PropTypes.any,
+      name: PropTypes.any,
```
There could be more specific `PropTypes` used than `any`.

lhellebr commented:
I, just as @kmalyjur , wasn't able to get the tooltips working out of the box => failed.

tlabaj committed:
Fixes #38879 - Update BarChart to Patternfly5

tlabaj commented:
@lhellebr @kmalyjur I addressed the comments.

The bar chart should show the tooltip on hover consistently now.

<img width="976" height="406" alt="image" src="https://github.com/user-attachments/assets/d285816c-d1f2-46fe-93fe-3b6747330294" />

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,173 @@
+const transformData = rawData => {
+  if (!rawData || !Array.isArray(rawData)) return [];
+
+  return rawData.map((item, index) => ({
```
Sorry, I know the `index` was in my suggestion before.

```suggestion
  return rawData.map(item => ({
```

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/BarChartHtmlTooltip.js:
```diff
@@ -0,0 +1,193 @@
+              className="bar-chart__tooltip-swatch"
+              style={{ '--bar-chart-legend-color': color }}
+            />
+            <span>Value:</span>
```
`Value` should be translated

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/BarChart.fixtures.js:
```diff
@@ -47,3 +47,14 @@ export const barChartConfigWithEmptyData = {
 };
```
Not used anymore

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,173 @@
+          dependentAxis
+          showGrid
+          label={yAxisLabel}
+          tickFormat={t => String(Math.round(t))}
```
When there is a max value of 1, the y-axis shows "0, 0, 1, 1, 1"

<img width="474" height="628" alt="Image" src="https://github.com/user-attachments/assets/212f2d54-17fc-419c-871c-1e5cf573968c" />

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,173 @@
+  }));
+};
+
+const CHART_WIDTH = 600;
+const CHART_HEIGHTS = {
+  small: 250,
+  medium: 280,
+  regular: 300,
+};
```
```suggestion
const CHART_WIDTH = 600;
const CHART_HEIGHTS = {
  small: 250,
  medium: 280,
  regular: 300,
};

const CHART_PADDING = {
  bottom: 60,
  left: 70,
  right: 20,
  top: 10,
};

const DOMAIN_PADDING = { x: [30, 25] };

const TOOLTIP = {
  width: 140,
  height: 70,
  padding: 10,
  pointerLength: 8,
};
```

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/BarChartHtmlTooltip.js:
```diff
@@ -0,0 +1,193 @@
+        width={tooltipWidth}
+        x={tooltipX}
+        y={tooltipY}
+      >
+        <div
+          className="bar-chart__tooltip"
+          xmlns="http://www.w3.org/1999/xhtml"
+        >
```
Adding role="tooltip" should help the screen readers

kmalyjur reviewed (changes requested):
Thank you, it looks great now!

However, in addition to the comments I'm leaving, I tried simplifying the tooltip positioning. I've put just these changes (it includes just two files, but the rest you have in this PR are needed) into this commit: https://github.com/theforeman/foreman/compare/develop...kmalyjur:foreman:10743
Could you please take a look and let me know what you think? It may not be perfect, but I just wanted to show the tooltip simplifying.

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -1,39 +1,173 @@
   noDataMsg,
   config,
   title,
-  unloadData,
```
I'm so sorry, I think I told you before that it's unused, which it is, but I forgot about: _"Received props should be kept the same, otherwise this task should also refactor the chart uses in the plugin."_
It should be there, and the default should be `false`.

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -16,24 +57,122 @@ const BarChart = ({
+    width: CHART_WIDTH,
+  });
+
+  const dimensions = getChartDimensions();
```
This is a nitpick, but it could be just:
```suggestion
  const dimensions = {
  height: CHART_HEIGHTS[config] || CHART_HEIGHTS.regular,
  width: CHART_WIDTH,
};
```

kmalyjur reviewed (approved):
Thank you, LGTM!

kmalyjur commented:
@lhellebr could you please verify?

lhellebr commented:
I know I may be very nitpicky for such a simple change... but this doesn't look good, and it was OK before:
<img width="798" height="545" alt="notgood" src="https://github.com/user-attachments/assets/8e7c5fcf-780a-498b-9a91-81b337fb768f" />

kmalyjur commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -58,6 +202,7 @@ BarChart.defaultProps = {
   unloadData: false,
   yAxisLabel: '',
   xAxisLabel: '',
+  yAxisLabelOffset: -12,
```
@tlabaj to @lhellebr's point, could the padding not be hardcoded?

kmalyjur replied:
Also, I'm not sure why the tests don't pass. Maybe rebasing could help?

tlabaj replied:
@lhellebr @kmalyjur I thought the ask was to add a prop so the offset could be adjusted if needed instead of hard coded.

kmalyjur replied:
I'm working on some changes regarding the offset.

lhellebr commented:
I've tested with packit build and it looks good, normally I would ACK but I'll have to retest this after the hardcoded offset is fixed

kmalyjur commented:
@tlabaj Following up on our last conversation, I was testing some edge cases with bigger numbers and made a few changes. It is not perfect, but I think it's usable.

Here is the updated file: https://gist.github.com/kmalyjur/46c07c9c46fcd07ca79668d8b9a14c64

TLDR of the changes:
- **Dynamic Padding**: Added a dynamic calculation for the left margin so it automatically scales. It's not ideal, when the numbers get too large, they overlap again, but I think it's a sufficient solution.
- **JS float bug**: Big numbers rendered as 6.00000001e27 and broke the left margin. It formats them now.
- **Tooltips**: Added a dynamic width calculation.

Please let me know what you think. Feel free to copy that updated file if you consider it good enough.

Max num length it works fine on:
<img width="914" height="692" alt="image" src="https://github.com/user-attachments/assets/4811d525-bd4c-4ac9-bab4-2af9bce85abe" />

When it's too long:
<img width="914" height="692" alt="image" src="https://github.com/user-attachments/assets/ff4819a5-01ed-43d9-8d69-0751bd45971c" />

Wider tooltip:
<img width="914" height="692" alt="image" src="https://github.com/user-attachments/assets/6c02931c-07ec-450f-bb7e-9702fb6a5071" />

kmalyjur commented:
Thanks, LGTM! The failing test does not seem related.
@lhellebr could you please verify, again? The main changes are ensuring that it can handle long values, you can see https://github.com/theforeman/foreman/pull/10743#issuecomment-4118048251.

lhellebr reviewed (approved):
Verified with packit build foreman-3.19.0-0.20260325143938836670.pr10743.9026.g4359cf7ed.el9.noarch , all of my findings have been addressed and it looks good. Thanks!

kmalyjur reviewed (approved):
Thank you @tlabaj and @lhellebr! 🥳

kmalyjur commented:
Rebase might fix the red tests?

kmalyjur commented:
Can be merged

ofedoren reviewed (approved):
Pro-forma ack, real acks are in the comments
